opal/mca/btl/sm/btl_sm_send.c

```diff
@@ -42,18 +43,25 @@ int mca_btl_sm_send (struct mca_btl_base_module_t *btl,
     mca_btl_sm_frag_t *frag = (mca_btl_sm_frag_t *) descriptor;
     const size_t total_size = frag->segments[0].seg_len;
 
-    /* in order to work around a long standing ob1 bug (see #3845) we have to always
-     * make the callback. once this is fixed in ob1 we can restore the code below. */
-    frag->base.des_flags |= MCA_BTL_DES_SEND_ALWAYS_CALLBACK;
+    if (frag->base.des_cbfunc) {
+        /* in order to work around a long standing ob1 bug (see #3845) we have to always
+         * make the callback. once this is fixed in ob1 we can restore the code below. */
+        frag->base.des_flags |= MCA_BTL_DES_SEND_ALWAYS_CALLBACK;
+    }
 
     /* header (+ optional inline data) */
     frag->hdr->len = total_size;
     /* type of message, pt-2-pt, one-sided, etc */
     frag->hdr->tag = tag;
 
+    /* clear the complete flag if it has been set */
+    frag->hdr->flags &= ~MCA_BTL_SM_FLAG_COMPLETE;
```

Comment:
The commit message doesn't mention this.  What is this change?

Reply:
Ah, yes. When using the AM-RDMA the frags are immediately re-used. btl/sm returns frags to the sender by setting that flag so it needs to be cleared before the frag can be reused. I can either clear it here or when the frag return is detected. I will add it to the commit message.

This is all about supporting the AM-RDMA usage.